include/libtorrent/disk_io_thread_pool.hpp

```diff
@@ -120,8 +124,6 @@ namespace libtorrent {
 		pool_thread_interface& m_thread_iface;
 
 		std::atomic<int> m_max_threads;
-		// the number of threads the reaper decided should exit
-		std::atomic<int> m_threads_to_exit;
```

Comment:
I don't think we can remove `m_threads_to_exit`. This PR tries to use `m_max_threads` instead but that breaks the scalable thread pool. At it is when the thread pool goes from max threads to idle it will reduce `m_max_threads` to zero which will prevent any new threads from being created. There needs to be a separate control for stopping threads due to lack of demand vs due to the configured max threads being reduced.

Reply:
yeah, this patch clearly doesn't work. Do you have any ideas of how to solve this in a simple way?

One other aspect is that when the session terminates immediately, the disk I/O threads don't come up until *after* `abort()` is called on `disk_io_thread`, so the destructor has to abort (and wait for the thread) again

Reply:
I was able to repro the memory leak. I found that, after removing the eager abort bail out in `thread_fun` the other issue which was preventing the queue from being processed was deferred job submission. The session can shutdown before jobs have been submitted in which case no thread is created to handle the deferred jobs. Simple solution is to always submit queued jobs just before initiating an abort. See #3744